### PR TITLE
Fix launch template reconciliation if bootstrap data secret cannot be read

### PR DESCRIPTION
**Pull request by AndiDog**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

When changing the machine pool's bootstrap config, e.g. `KubeadmConfig/{a => b}`, which is supported after CAPI's fix https://github.com/kubernetes-sigs/cluster-api/pull/8667, CAPA had a bug where it would continue even after the new bootstrap secret cannot be read, for instance because it doesn't exist yet as the bootstrap provider still needs a few seconds to make it ready. That could lead to a catastrophic misconfiguration where the launch template's user data is empty, typically leading to new instances not getting bootstrapped at all. Detailed problem description in [this comment](https://github.com/kubernetes-sigs/cluster-api/issues/8858#issuecomment-1737930898).

This provides the obvious fix. I couldn't find a great place to test this, since `ReconcileLaunchTemplate` is mocked away in tests.

I tested this successfully by changing the bootstrap config reference and observing how CAPI sets the new data secret name, CAPA intermittently failing to read the new secret until it's ready, and only then CAPA writing a new launch template version with the new bootstrap secret's data.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Part of https://github.com/kubernetes-sigs/cluster-api/issues/8858, but not a fix. Even with this patch, changing the bootstrap config object doesn't lead to an instance refresh (rolling out new nodes) because CAPA doesn't consider it a change. We need to discuss further how to solve the main issue.

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Fix launch template reconciliation if bootstrap data secret cannot be read
```

### CODE REVIEW

File: pkg/cloud/services/ec2/launchtemplate.go
Comment (richardcase) [inline pkg/cloud/services/ec2/launchtemplate.go:62]: Its probably worth logging as well here.

Comment (AndiDog) [reply pkg/cloud/services/ec2/launchtemplate.go:62]: In the whole function, errors are only returned, not logged. The callers then log it like this:

```
if err := ec2Svc.ReconcileLaunchTemplate(machinePoolScope, canUpdateLaunchTemplate, runPostLaunchTemplateUpdateOperation); err != nil {
	r.Recorder.Eventf(machinePoolScope.AWSMachinePool, corev1.EventTypeWarning, "FailedLaunchTemplateReconcile", "Failed to reconcile launch template: %v", err)
	machinePoolScope.Error(err, "failed to reconcile launch template")
	return err
}
```

I prefer keeping it this way since we don't get two log lines about the same error.

Comment (richardcase) [reply pkg/cloud/services/ec2/launchtemplate.go:62]: Sounds good to me.

**Review by cnmcavoy — APPROVED** [review]
/lgtm

**Comment by AndiDog** [comment]
@richardcase The test `pull-cluster-api-provider-aws-e2e-eks` was waiting for an EKS cluster to be created and the result here in GitHub says "Pod got deleted unexpectedly". Do you have an idea for this issue? Can it be because of https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4575 not merged into my PR?

**Comment by AndiDog** [comment]
Trying the test again after having merged `main` (incl. possible test fix https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4575)

/test pull-cluster-api-provider-aws-e2e-eks

**Comment by AndiDog** [comment]
@cnmcavoy @faiq I merged `main` and the E2E test worked this time. Switched to rebase since merge isn't allowed. Need another LGTM – no changes happened to my diff.

**Review by faiq — APPROVED** [review]

**Pull request merged**
